tutorial/20_recipes/013_wilcoxon_pruner.py

```diff
@@ -0,0 +1,324 @@
+            if remaining_patience == 0:
+                idxs[:] = best_idxs
+                cost = best_cost
+                remaining_patience = options.patience
+
+    return best_idxs
+
+
+###################################################################################################
+# We make a random dataset of TSP.
+
+
+def make_dataset(num_vertex: int, num_problem: int, seed: int = 0) -> np.ndarray:
+    rng = np.random.default_rng(seed=seed)
+    return rng.random((num_problem, num_vertex, 2))
+
+
+dataset = make_dataset(
+    num_vertex=100,
+    num_problem=50,
+)
+
+N_TRIALS = 50
+
+
+###################################################################################################
+# We set a very small number of SA iterations for demonstration purpose.
+# In practice, you should set a larger number of iterations.
+
+
+N_SA_ITER = 10000
+count = 0
+
+
+###################################################################################################
+# We counts the number of evaluation to know how many problems is pruned.
+
+
+num_evaluation = 0
+
+
+###################################################################################################
+# In this tutorial, we optimize three parameters: ``T0``, ``alpha``, and ``patience``.
+#
+# ``T0`` and ``alpha`` defining the temperature schedule
+# ---------------------------------------------------------------------------------------
+#
+# In simulated annealing, it is important to determine a good temperature scheduling, but
+# there is no "silver schedule" that is good for all problems, so we must tune the schedule
+# for this problem.
+# This code parametrizes the temperature as a monomial function ``T0 * (1 - t) ** alpha``, where
+# `t` progresses from 0 to 1. We try to optimize the two parameters ``T0`` and ``alpha``.
+#
+# ``patience``
+# -----------------------------
+#
+# This parameter specifies a threshold of how many iterations we allow the annealing process
+# continue without updating the best value. Practically, simulated annealing often drives
+# the solution far away from the current best solution, and rolling back to the best solution
+# periodically often improves optimization efficiency a lot. However, if the rollback happens
+# too often, the optimization may get stuck in a local optimum, so we must tune the threshold
+# to a sensible amount.
+#
+# .. note::
+#     Some samplers, including the default ``TPESampler``, currently cannot utilize the
+#     information of pruned trials effectively (especially when the last intermediate value
+#     is not the best approximation to the final objective function).
+#     As a workaround for this issue, you can return an estimation of the final value
+#     (e.g., the average of all evaluated values) when ``trial.should_prune()`` returns ``True``,
+#     instead of `raise optuna.TrialPruned()`.
+#     This will improve the sampler performance.
+
+
+def objective(trial: optuna.Trial) -> float:
```

Comment:
It would be great to explain this part. For example, `We define the objective function to be optimized as follows. We early stop the evaluation by using the pruner.`.